### ISSUE TITLE
Bug of tf.data.TFRecordDataset? Couldn't use tf.reshape after the operations of tf.data.TFRecordDataset

### ISSUE THREAD

**Issue by zhongxing9006** [tensorflow/tensorflow#16313]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4.1
- **Python version**: 2.7.12
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 8.0/6.0
- **GPU model and memory**: Nvidia GeForce GTX TITAN X 12GB
- **Exact command to reproduce**:

### Describe the problem
I want to use  function **tf.profiler.ProfileOptionBuilder.float_operation** to show the flops of the model. But it need a certain input shape while the the output shape of **tf.data.TFrecordDataset** is like (?, 32,32,3). When I want to use tf.reshape to reshape the output of **tf.data.TFrecordDataset**, it generates an error "Input to reshape is a tensor with 64512 values, but the requested shape has 98304".  

### Source code / logs
def dataset_input(self,dataset_type):
     with tf.variable_scope("batch_"+dataset_type):
            def parser(record):
                features = tf.parse_single_example(
                    record,
                    features={
                        'image': tf.FixedLenFeature([], tf.string),
                        'label': tf.FixedLenFeature([], tf.int64)
                    })
                image, label = features['image'], features['label']
                height, width, channels = self.input_size, self.input_size, self.input_dim
                image = tf.decode_raw(image, tf.uint8)
                image = tf.reshape(image, [height, width, channels])
                return image,label

            dataset = tf.data.TFRecordDataset([self.dataset_dir[dataset_type]])
            dataset = dataset.map(parser)
            dataset = dataset.shuffle(buffer_size=50000)
            dataset = dataset.batch(self.batch_size)
            dataset = dataset.repeat()
            iterator = dataset.make_one_shot_iterator()
            features, labels = iterator.get_next()
            **features = tf.reshape(features, [self.batch_size, self.input_size, self.input_size, self.input_dim])**
            return features, labels

**Comment by yaox12** [comment]
Did you try to reshape in `parser`?

**Comment by monomon** [comment]
Looks like simply your data is smaller than what you are trying to resize to. Are you sure the number of channels (`input_dim`) is correct, and image is not e.g. with an alpha channel too?
How are you generating the dataset?

This discussion is better moved to stack overflow, since it does not look like a bug in tf.

**Comment by mrry** [comment]
The most likely cause is that the input file does not contain an exact multiple of `self.batch_size` records, and your program is crashing when the final batch (containing `total_batch_size * self.batch_size` rows) is emitted. To avoid this problem, avoid using `self.batch_size` in your model, for example by replacing the `tf.reshape()` call with:

```python
features = tf.reshape(features, [-1, self.input_size, self.input_size, self.input_dim])
```

**Comment by zhongxing9006** [comment]
@monomon  I have already verified the value of input_dim and it is correct. I also use tf.TFRecordReader to read the dataset directly and it worked well.
    def loadDataFromRecord(self, dataset_label, name = None):
        with tf.variable_scope("batch_"+name):
            reader = tf.TFRecordReader()
            dataset_dir = [self.dataset_dir[dataset_label]]
            filename_queue = tf.train.string_input_producer(dataset_dir, shuffle=True, capacity=64)
            _, serializedData = reader.read(filename_queue)

            features = tf.parse_single_example(
                serializedData,
                features={
                    'image': tf.FixedLenFeature([], tf.string),
                    'label': tf.FixedLenFeature([], tf.int64) 
                })
            image, label = features['image'], features['label']
            height, width, channels = self.input_size, self.input_size, self.input_dim
            image = tf.decode_raw(image, tf.uint8)
            image = tf.reshape(image, [height, width, channels])
            min_after_dequeue = 10000
            capacity = min_after_dequeue + 3*self.batch_size
            image_batch, label_batch = tf.train.shuffle_batch(
                [image, label], batch_size=self.batch_size,
                capacity=capacity, min_after_dequeue=min_after_dequeue)
            image_batch = tf.reshape(image_batch, [self.batch_size, self.input_size, self.input_size, self.input_dim])
            return image_batch, label_batch

**Comment by zhongxing9006** [comment]
I solve this problem.  The total number of samples is not an integral multiple of the number of batch.
